gtn/functions/compose.cpp

```diff
@@ -450,3 +456,931 @@ Graph compose(
+      // epsilon
+      if (idx < toExploreNumArcs[i].second) {
+        checkEpsilonArcPair.second = true;
+      }
+
+      break;
+    }
+  }
+
+  return std::make_tuple(isValid, nodePair, arcPair, checkEpsilonArcPair);
+}
+
+// Takes a pair of nodes, where each member of pair comes from a different
+// graph and calculate a vector of number of arcs in the cross product of
+// arcs outgoing from each pair.
+// This should be a kernel call
+std::tuple<std::vector<int>, std::vector<std::pair<int, int>>>
+calculateArcCrossProductOffset(
+    const std::vector<std::pair<int, int>>& toExploreNodePair,
+    const GraphDataParallel& graphDP1,
+    const GraphDataParallel& graphDP2,
+    bool inOrOutArc) {
+  std::vector<std::pair<int, int>> toExploreNumArcs(toExploreNodePair.size());
+  std::vector<int> arcCrossProductOffset(toExploreNodePair.size());
+
+  // No dependence between iterations
+  for (size_t i = 0; i < toExploreNodePair.size(); ++i) {
+    int node = toExploreNodePair[i].first;
+    // Special case if it is the last node. Then the offset becomes
+    // the number of arcs
+    int inArcOffset = ((node + 1) == graphDP1.inArcOffset.size())
+        ? graphDP1.inArcs.size()
+        : graphDP1.inArcOffset[node + 1];
+    int outArcOffset = ((node + 1) == graphDP1.outArcOffset.size())
+        ? graphDP1.outArcs.size()
+        : graphDP1.outArcOffset[node + 1];
+
+    const int numArcsFirst = inOrOutArc
+        ? inArcOffset - graphDP1.inArcOffset[node]
+        : outArcOffset - graphDP1.outArcOffset[node];
+
+    node = toExploreNodePair[i].second;
+    // Special case if it is the last node. Then the offset becomes
+    // the number of arcs
+    inArcOffset = ((node + 1) == graphDP2.inArcOffset.size())
+        ? graphDP2.inArcs.size()
+        : graphDP2.inArcOffset[node + 1];
+    outArcOffset = ((node + 1) == graphDP2.outArcOffset.size())
+        ? graphDP2.outArcs.size()
+        : graphDP2.outArcOffset[node + 1];
+
+    const int numArcsSecond = inOrOutArc
+        ? inArcOffset - graphDP2.inArcOffset[node]
+        : outArcOffset - graphDP2.outArcOffset[node];
+
+    toExploreNumArcs[i] = std::make_pair(numArcsFirst, numArcsSecond);
+    arcCrossProductOffset[i] = numArcsFirst * numArcsSecond;
+  }
+
+  return std::make_tuple(arcCrossProductOffset, toExploreNumArcs);
+}
+
+// This function needs to be thread safe since multiple threads can
+// can call it and they will overlap on curIdx and dstIdx
+void calculateNumArcsAndNodesToExplore(
+    int curIdx,
+    int dstIdx,
+    const std::vector<bool>& reachable,
+    std::vector<bool>& newNodes,
+    std::vector<bool>& toExplore,
+    std::vector<int> numOutArcs,
+    std::vector<int> numInArcs) {
+  if (reachable[dstIdx]) {
+    // Atomic test and set for newNodes
+    if (!newNodes[dstIdx]) {
+      newNodes[dstIdx] = true;
+      toExplore[dstIdx] = true;
+    }
+
+    // These are atomic increments
+    numOutArcs[curIdx]++;
```

Comment:
`numOutArcs` and `numInArc` are passed **by value**. The two increments
here are visible only inside the function.

Reply:
Thanks!! will fix